File: ***Questions-Solutions***/Two-Dimensional-Arrays/Python/Q03Q04RotateMatrix.py
import copy
import queue
import math

# Implement your function below.
# n = # rows = # columns in the given 2d array
def rotate(given_array, n):
    # rotated = copy.deepcopy(given_array)
    # NOTE: To solve it in place, write this function so that you
    # won't have to create rotated.
    temp_items = [-1]*4
    for i in range(math.ceil(n/2)):
        for j in range(math.floor(n/2)):
            (current_i,current_j) = (i,j)
            for t in range(4):
                temp_items[t] = given_array[current_i][current_j]
                (current_i,current_j) = sub_rotate(current_i,current_j,n)
            for t in range(4):
                given_array[current_i][current_j] = temp_items[(t+3)%4]
                (current_i,current_j) = sub_rotate(current_i,current_j,n)
    return given_array



def sub_rotate(i,j,n):
    return j,n-1-i
                    
# NOTE: Feel free to use the following function for testing.
# It converts a 2-dimensional array (a list of lists) into
# an easy-to-read string format.
def to_string(given_array):
    list_rows = []
    for row in given_array:
        list_rows.append(str(row))
    return '[' + ',\n '.join(list_rows) + ']'



# NOTE: The following input values will be used for testing your solution.
a1 = [[1, 2, 3],
      [4, 5, 6],
      [7, 8, 9]]
print(to_string(rotate(a1, 3))) #should return:
# [[7, 4, 1],
#  [8, 5, 2],
#  [9, 6, 3]]

a2 = [[1, 2, 3, 4],
      [5, 6, 7, 8],
      [9, 10, 11, 12],
      [13, 14, 15, 16]]
print(to_string(rotate(a2, 4))) #should return:
# [[13, 9, 5, 1],
#  [14, 10, 6, 2],
#  [15, 11, 7, 3],
#  [16, 12, 8, 4]]

File: ***Questions-Solutions***/Two-Dimensional-Arrays/C++/Q03Q04RotateMatrix.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cmath>

struct pair{
    int newi;
    int newj;
};

pair rotate_sub(int i,int j,int n){
    pair out;
    out.newi = j;
    out.newj = n-1-i;
    return out;
}

void rotate(std::vector<std::vector<int>> ar,int n){
    int current_i;
    int current_j;
    pair p;
    for(int i=0;i<ceil((float)n/2);i++){
        for(int j=0;j<floor((float)n/2);j++){
            std::vector<int> tmp{-1,-1,-1,-1};
            current_i = i;
            current_j = j;
            for(int k=0;k<4;k++){
                tmp[k] = ar[current_i][current_j];
                p=rotate_sub(current_i,current_j,n);
                current_i = p.newi;
                current_j = p.newj;
            }
            for(int k=0;k<4;k++){
                ar[current_i][current_j] = tmp[(k+3)%4];
                p = rotate_sub(current_i,current_j,n);
                current_i = p.newi;
                current_j = p.newj;
            }
        }
    }
    
    for(int i=0;i<n;i++){
        for(int j=0;j<n;j++){
            std::cout<<ar[i][j]<<" ";
        }
        std::cout<<"\n";
    }
}

int main(){
    
    std::vector<std::vector<int>> ar{{1,2,3,4,5},
        {6,7,8,9,1},
        {2,3,4,5,6},
        {7,8,9,1,2},
        {3,4,5,6,7}};
    rotate(ar,5);
    
    return 0;
}
